Create Azure node from Compute Gallery image, OS disk delete

rgharris opened the pull request:
## Create Azure node from Compute Gallery image, OS disk delete

### Description

Source issues:
- #1643
- #1644

[Azure Compute Gallery](https://docs.microsoft.com/en-us/azure/virtual-machines/shared-image-galleries) is the recommended way to create custom images for Azure VMs. Supporting these works just like the `AzureImage` in libcloud today except you provide a single `id` under `imageReference` (instead of multiple key/values). See this [example request](https://docs.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update#create-a-vm-from-a-custom-image) and the [corresponding spec](https://docs.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update#imagereference). 

Azure VM OS disks can be configured at the time of VM creation to be deleted when the VM is deleted. It would be helpful if libcloud supported this option. There is not an option in libcloud to do this on delete and it is convenient in some use cases to delete the disk regardless of the method (CLI, portal, libcloud, etc.). Supporting this requires a `deleteOption` of `Delete` under `OSDisk` when creating a VM. See the [OSDisk spec](https://docs.microsoft.com/en-us/rest/api/compute/virtual-machines/create-or-update#osdisk) for details. 

A newer `api-version` is required for `create_node` (`deleteOption` not accepted otherwise).

### Status

Done, ready for review. 

The need to change the `api-version` for the Azure create node API may be problematic. I have tested `deleteOption` with Azure VMs using Compute Gallery Images with user code. I have added simple tests as well.

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

bverschueren committed:
add support for retrieving private key from JSON object/string/file

bverschueren committed:
add more thorough verification of PEM key

bverschueren committed:
update docs with service account credentials directly as dict

Commit:
Add ScalewayStorageDriver, ScalewayConnectionAWS4 and correspoding regions

Commit:
Add missing types

Commit:
Add simple test for scalewa storage driver

Commit:
Add documentation and example based on the Minio docs

Commit:
Add corrections. Use region concept from S3 driver.

Commit:
SCW storage driver: Update tests to be more S3 compliant

Commit:
Fix typo

Commit:
Use NotImplementedError instead

Commit:
Improve test case

murlock committed:
Fix SignedHeaders

Without this patch:
Signed headers: accept-encoding;content-type;host;user-agent;x-amz-content-sha256;x-amz-date;x-amz-acl;x-amz-meta-created;x-amz-meta-owner;x-amz-storage-class

With this patch:
Signed headers: accept-encoding;content-type;host;user-agent;x-amz-acl;x-amz-content-sha256;x-amz-date;x-amz-meta-created;x-amz-meta-owner;x-amz-storage-class

The sort was done on headers with upper/lower case, this patch fix this
beahiavor.

reixd committed:
Merge pull request #1 from murlock/scw_object_storage_driver

Fix SignedHeaders for Scaleway Provider

Commit:
In order to download (private) objects we need to be able to generate a valid CDN URL,
hence shamefully just use the working code from the S3StorageDriver.
Also update tests.

Kami committed:
Merge branch 'scw_object_storage_driver' of https://github.com/reixd/libcloud into reixd-scw_object_storage_driver

Commit:
Add corrections

Commit:
Linting and black

mazerty committed:
fix #1641: [EC2] Add missing "creation_date" NodeImage extra

Commit:
Create from Compute Gallery image, OS disk delete
- New api-version required for create_node
- Closes apache#1643
- Closes apache#1644

micafer committed:
fist solution for #1646

micafer committed:
Add microversion support

micafer committed:
Update docs

micafer committed:
Remove change from other issue

micafer committed:
Fix #1649

micafer committed:
Fix #1650

micafer committed:
Fix typo

micafer committed:
Fix black test

micafer committed:
Fix black test

micafer committed:
Merge branch 'fix_1650' of github.com:micafer/libcloud into ost_microversion

micafer committed:
Fix style

micafer committed:
Fix typo

micafer committed:
Merge branch 'ost_microversion' of github.com:micafer/libcloud into ost_microversion

micafer committed:
Fix error

Kami committed:
Add test which verifies the code doesn't throw if creationDate attribute
is not present.

Kami commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -703,7 +739,7 @@ def create_node(
 
         r = self.connection.request(
             target,
-            params={"api-version": RESOURCE_API_VERSION},
```
As you pointed out in the PR description, changing this version may be problematic.

To be on the safe side, I think it would be good to do the following:

1. Add new ``api_version`` to the driver constructor and allow user to override default API version (we already use similar pattern in many other drives).
2. Document this change in the upgrade_nodes.rst and show an example how user can utilize ``api_version`` constructor argument to set it to the previous default value.

rgharris replied:
Thanks for the feedback @Kami. 

There are various hardcoded `api-version` values (e.g., "2019-06-01") in the driver in addition to many references to `RESOURCE_API_VERSION`.

Which solution would fit best with the existing patterns for changes like this?

1. Leave the existing hardcoded version values. Update all using `RESOURCE_API_VERSION` to use the updated value via `self.api_version` (allow override).
2. Update all using `RESOURCE_API_VERSION` or hardcoded version values to use the updated value via `self.api_version` (allow override).
3. Move just `create_node` to use the new API version and add an extra parameter allowing the version to be overridden. This doesn't align with your suggestion of course but does minimize the potential impact to other methods.

1 & 2 do cause some potential issues: we will be affecting many methods, not just `create_node` that I have modified in this pull request. Updating the version for many methods is likely helpful as the current default is an older preview version. But further testing would be needed.

Kami replied:
We could start by trying with updating all the places to use the latest version and what's specified via ``api_version`` argument and then go from there - depending on if end to end tests pass or not. So in short, what you proposed via 2).

Of course end to end tests do not cover every single possible scenario and setup, but it should be a start and give us at least some basic confidence.

rgharris replied:
I gave that a try. It looks like Azure has different supported versions per service. For example, `2021-07-01` is supported for [VMs](https://docs.microsoft.com/en-us/rest/api/compute/virtual-machines) but not [disks](https://docs.microsoft.com/en-us/rest/api/compute/disks). [Public IP addresses](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/public-ip-addresses) have differing versions as well. 

So, using a common API version across all requests will not work. This does explain why newer code has used hardcoded version numbers. Some variation of solution (3) will likely be necessary. Or we could do a deeper dive into differences in the API versions and move to the new version with no override (for `create_node` and optionally other VM APIs) if no breaking changes. What are your thoughts on those options? 

An error when using `2021-07-01` for `ex_get_volume`:
> [NoRegisteredProviderFound] No registered resource provider found for location 'westus2' and API version '2021-07-01' for type 'disks'. The supported api-versions are '2016-04-30-preview, 2017-03-30, 2018-04-01, 2018-06-01, 2018-09-30, 2019-03-01, 2019-07-01, 2019-11-01, 2020-05-01, 2020-06-30, 2020-09-30, 2020-12-01, 2021-04-01, 2021-08-01, 2021-12-01'. The supported locations are 'southeastasia, eastus2, centralus, westeurope, eastus, northcentralus, southcentralus, westus, northeurope, eastasia, brazilsouth, westus2, westcentralus, ukwest, uksouth, japaneast, japanwest, canadacentral, canadaeast, centralindia, southindia, australiaeast, australiasoutheast, koreacentral, westindia, francecentral, southafricanorth, uaenorth, australiacentral, switzerlandnorth, germanywestcentral, norwayeast, jioindiawest, westus3, swedencentral'.

Kami replied:
Thanks for checking.

In this case, it sounds like it may be the best to only update api version used with the ``create_node()`` method.

And to make code a bit easier to maintain going forward (and in the case user needs to override the API version for some reason), we should probably define those versions as part of module level constants - something along the lines of ``VMS_API_VERSION, DISKS_API_VERSION``, etc.

If you want to try updating versions for other resources as well, that's fine, but it will likely require more work testing everything.

rgharris replied:
In 83e6f51d03ad9010093f44b29491a84cabb09858 I added constants per resource type API version. This required a few methods to use updated API versions due to inconsistencies within some resources. I have done some minimal testing of all methods. I would be okay with using this pattern or just adjusting `create_node()` to simplify testing, let me know what you think.

These had version changes in the commit:
```
# updated to latest version due to changes in create_node
list_nodes 
create_node 
reboot_node
destroy_node 
ex_get_node
start_node
stop_node  

# updated to version used by ex_resize_volume
create_volume 
list_volumes
ex_get_volume 

# updated to version used by ex_delete_public_ip
ex_get_public_ip 
ex_list_public_ips
ex_create_public_ip  

# updated to version used by ex_update_nic_properties
ex_create_network_interface
```

rgharris replied:
In c3b12ecb20872ed5b0af9391d4dc539f017c4844 I have reverted the version number change for the volume methods due to `properties.accountType` in the request changing to `sku.name` in the newer version. Found this after further manual testing. I've added a comment noting that ideally down the road all volume operations would be updated to use the same version. 

@Kami do you think this PR is ready to be merged? If not, any suggestions on this API version consolidation?

Updated list of version changes:
```
# updated to latest version due to changes in create_node
list_nodes 
create_node 
reboot_node
destroy_node 
ex_get_node
start_node
stop_node 

# updated to version used by ex_delete_public_ip
ex_get_public_ip 
ex_list_public_ips
ex_create_public_ip  

# updated to version used by ex_update_nic_properties
ex_create_network_interface
```

Kami commented:
Thanks for the contribution.

I added a comment on how we could make the API version change a bit safer in case that change brakes the behavior for some user / setup, besides that, LGTM.

Kami committed:
Merge branch 'scw_object_storage_driver' of https://github.com/reixd/libcloud into reixd-scw_object_storage_driver

Kami committed:
Add changelog entry.

Kami committed:
Re-generate supported providers tables.

Kami committed:
Merge branch 'reixd-scw_object_storage_driver' into trunk

Kami committed:
Merge branch 'LIBCLOUD-996_gcp_svc_acct_credentials_from_json' of https://github.com/bverschueren/libcloud into bverschueren-LIBCLOUD-996_gcp_svc_acct_credentials_from_json

Kami committed:
Reformat code with black.

Kami committed:
Add changelog entry.

Kami committed:
Fix formatting.

Kami committed:
Fix black tox target, make sure we also run it on code in libcloud/
directory.

Kami committed:
Reformat code with black.

Kami committed:
Add a test case for scenario where the PEM key is invalid.

Kami committed:
Add new test fixture files.

Kami committed:
Upgrade black to the latest stable version.

Kami committed:
Reformat code with the latest version of black.

Kami committed:
Merge pull request #1653 from Kami/upgrade_black

Upgrade black to the latest stable version

micafer committed:
Add test

micafer committed:
Update changelog

micafer committed:
Merge branch 'trunk' into fix_1649

micafer committed:
Add test and update changelog

micafer committed:
Merge branch 'trunk' into fix_1650

micafer committed:
Move microvertion to connection

micafer committed:
Improve test

micafer committed:
Merge pull request #1652 from micafer/fix_1650

headers field are overwrited in case of POST of PUT methods in OpenStack connection

micafer committed:
Merge branch 'trunk' into fix_1649

micafer committed:
Merge branch 'trunk' into ost_microversion

micafer committed:
Merge pull request #1651 from micafer/fix_1649

Personality field in the server requests of OpenStack must be optional

micafer committed:
Fix #1646

micafer committed:
Add max_results param

micafer committed:
Use paginated_request if ex_project is None

r2ronoha committed:
Add support for eu-south-1 in EC2

r2ronoha committed:
Merge branch 'trunk' into ec2_eu-south-1_v3.3.1

r2ronoha committed:
quites

r2ronoha committed:
quotes

micafer committed:
Add test

micafer committed:
Update changelog

r2ronoha committed:
VCloud 5.5. accept version

r2ronoha committed:
tests

Commit:
Extract API versions per resource
- Move to latest version currently used per resource
- Minimize unnecessary version upgrades
- VM_API_VERSION upgraded to latest due to new functionality

Commit:
Revert consolidation of disk API versions
- Creation now requires `sku.name` instead of `properties.accountType`
- Merge DISK_RESIZE_API_VERSION  into DISK_API_VERSION in future

Kami committed:
Scrape, re-generate and re-format EC2 sizes and prices data.

Kami committed:
Revert "VCloud 5.5. accept version"

This reverts commit 06f5c4d938484033b9b20e5116163fa6cb4a853b.

Kami committed:
Add changelog entry.

Kami committed:
Remove duplicated entry.

Kami committed:
Merge pull request #1656 from r2ronoha/ec2_eu-south-1_v3.3.1

Add support for eu-south-1 in EC2 and VCloud accept version

Kami committed:
Merge pull request #1648 from micafer/ost_microversion

Enable to set microversion in OpenStack driver

Kami committed:
Add changelog entry for #1648.

Kami committed:
Re-generate supported providers tables.

Kami committed:
Merge branch 'fix_1646' of https://github.com/micafer/libcloud into micafer-fix_1646

Kami committed:
Update changelog entry.

Kami committed:
Add back a safe guard so the code doesn't throw in case the response
list item desn't contain "items" entry.

Kami committed:
Update changelog entry.

Kami committed:
Merge pull request #1655 from micafer/fix_1646

Add paginated request method to GCEConnection

Kami reviewed (approved):
Sorry for the delay, LGTM - I will go ahead and merge the changes.

Kami committed:
Merge branch 'azure-arm-compute-gallery-and-os-disk-delete' of https://github.com/rgharris/libcloud into rgharris-azure-arm-compute-gallery-and-os-disk-delete

Kami committed:
Fix lint.

Kami committed:
Reformat code with black.
